siliconcompiler/schema/schema_obj.py

```diff
@@ -194,204 +229,162 @@ def record_history(self):
+
+        The normalization rules are as follows:
+        - If a scalar is provided for a list type, it is promoted to a list of
+        one element.
+        - If a list is provided for a tuple type, it is cast to a tuple (since
+        the JSON module serializes tuples as arrays, which are deserialized into
+        lists).
+        - Elements inside lists and tuples are normalized recursively.
+        - All non-list values have a string representation that gets cast to a
+        native Python type (since we receive strings from the CLI):
+          - bool: accepts "true" or "false"
+          - ints and floats: cast as if by int() or float()
+          - tuples: accepts comma-separated values surrounded by parens
         '''
-        Internal recursive function that searches the Chip schema for a
-        match to the combination of *args and fields supplied. The function is
-        used to set and get data within the dictionary.
 
-        Args:
-            cfg(dict): The cfg schema to search
-            keypath (str): Concatenated keypath used for error logging.
-            args (str): Keypath/value variable list used for access
-            field(str): Leaf cell field to access.
-            mode(str): Action (set/get/add/getkeys/getkeys)
-            clobber(bool): Specifies to clobber (for set action)
+        if value is None and not Schema._is_list(field, sc_type):
+            # None is legal for all scalars, but not within collection types
+            # TODO: could consider normalizing "None" for lists to empty list?
+            return value
+
+        if field in ('value', 'defvalue'):
+            try:
+                return Schema._normalize_value(value, sc_type)
+            except TypeError:
+                # Re-raise exception with a consistent message
+                raise TypeError(f'Invalid value {value}: expected {sc_type}') from None
+        else:
+            return Schema._normalize_field(value, sc_type, field)
 
-        '''
+    @staticmethod
+    def _normalize_value(value, sc_type):
+        if sc_type.startswith('['):
+            if not isinstance(value, list):
+                value = [value]
+            base_type = sc_type[1:-1]
+            return [Schema._normalize_value(v, base_type) for v in value]
+
+        if sc_type.startswith('('):
+            # TODO: make parsing more robust to support tuples-of-tuples
+            if isinstance(value, str):
+                value = value[1:-1].split(',')
+            elif not (isinstance(value, tuple) or isinstance(value, list)):
+                raise TypeError
+
+            base_types = sc_type[1:-1].split(',')
+            if len(value) != len(base_types):
+                raise TypeError
+            return tuple(Schema._normalize_value(v, base_type) for v, base_type in zip(value, base_types))
+
+        if sc_type == 'bool':
+            if value == 'true': return True
+            if value == 'false': return False
```

Comment:
`if value in ('true', 'True')`? Ditto for false?

Reply:
I'm thinking of keeping as-is to maintain parity with our current API, which only supports lower-case "true"/"false" for strings. However, I'd be open to changing this if others think it's a good idea.